### PR TITLE
[ESIMD] Enable -fsycl-esimd-force-stateless-mem by default

### PR DESCRIPTION
**Pull request by v-klochkov**

### CODE REVIEW

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 1]
```diff
@@ -5390,9 +5394,10 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
       for (auto &Macro : D.getSYCLTargetMacroArgs())
         CmdArgs.push_back(Args.MakeArgString(Macro));
     }
-    if (Args.hasFlag(options::OPT_fsycl_esimd_force_stateless_mem,
-                     options::OPT_fno_sycl_esimd_force_stateless_mem, false))
-      CmdArgs.push_back("-fsycl-esimd-force-stateless-mem");
+    if (IsSYCL &&
+        !Args.hasFlag(options::OPT_fsycl_esimd_force_stateless_mem,
+                      options::OPT_fno_sycl_esimd_force_stateless_mem, true))
+      CmdArgs.push_back("-fno-sycl-esimd-force-stateless-mem");
```
Comment (mdtoguchi) [inline clang/lib/Driver/ToolChains/Clang.cpp:5400]: This looks like it is being added twice?

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 2]
```diff
@@ -5390,9 +5394,10 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
       for (auto &Macro : D.getSYCLTargetMacroArgs())
         CmdArgs.push_back(Args.MakeArgString(Macro));
     }
-    if (Args.hasFlag(options::OPT_fsycl_esimd_force_stateless_mem,
-                     options::OPT_fno_sycl_esimd_force_stateless_mem, false))
-      CmdArgs.push_back("-fsycl-esimd-force-stateless-mem");
+    if (IsSYCL &&
```
Comment (mdtoguchi) [inline clang/lib/Driver/ToolChains/Clang.cpp:5397]: `IsSYCL` should already be true at this point.

File: clang/test/Preprocessor/sycl-esimd-force-stateless-mem.cpp [hunk 3]
```diff
@@ -1,11 +1,13 @@
 /// This test checks that the macro __ESIMD_FORCE_STATELESS_MEM is automatically
-/// defined only if the option -fsycl-esimd-force-stateless-mem is used.
+/// defined by default with -fsycl, -fsycl-is-device or -fsycl-is-host.
```
Comment (premanandrao) [inline clang/test/Preprocessor/sycl-esimd-force-stateless-mem.cpp:2]:
```suggestion
/// defined by default with -fsycl-is-device or -fsycl-is-host.
```
Just a nit.   *This* test being cc1 does not test what happens with -fsycl.

Comment (v-klochkov) [reply clang/test/Preprocessor/sycl-esimd-force-stateless-mem.cpp:2]: This test was inspired by an existing test for SYCL: https://github.com/intel/llvm/blob/sycl/clang/test/Preprocessor/sycl-macro.cpp
cc1 needs to be only checked with -fsycl-is-device and -fsycl-is-host. -fsycl option seem not relevant to cc1.

**Review by mdtoguchi — APPROVED** [review]

**Review by premanandrao — APPROVED** [review]
LGTM

**Comment by mmeterel** [comment]
@v-klochkov Is this going to be part of 2024.1 release?

**Comment by v-klochkov** [comment]
> @v-klochkov Is this going to be part of 2024.1 release?

It is too late for 2024.1, it will be in 2024.2

**Pull request merged**